source/util/parse_number.h

```diff
@@ -0,0 +1,143 @@
+enum class EncodeNumberStatus {
+  kSuccess = 0,
+  // Unsupported bit width etc.
+  kUnsupported,
+  // Expected type (IdType) is not a scalar int or float, or putting a
+  // negative number in an unsigned literal.
+  kInvalidUsage,
+  // Number value does not fit the bit width of the expected type etc.
+  kInvalidText,
+};
+
+// Parses a numeric value of a given type from the given text.  The number
+// should take up the entire string, and should be within bounds for the
+// target type.  On success, returns true and populates the object
+// referenced by value_pointer. On failure, returns false.
+template <typename T>
+bool ParseNumber(const char* text, T* value_pointer);
+
+// Parses an integer value of a given |type| from the given |text| and encodes
+// the number by the given |emit| function. On success, returns
+// EncodeNumberStatus::kSuccess and an empty error message string, the parsed
+// number will be consumed by the given |emit| function word by word (least
+// significant word first). On failure, this function returns the error
+// encoding status and error message string to describe the error, |emit|
+// function will not be called on failure. Integers up to 64 bits are
+// supported.
+std::tuple<EncodeNumberStatus, std::string> ParseAndEncodeIntegerNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+// Parses a floating point value of a given |type| from the given |text| and
+// encodes the number by the given |emit| funciton. On success, returns
+// EncodeNumberStatus::kSuccess and an empty error message string, the parsed
+// number will be consumed by the given |emit| function word by word (least
+// significant word first). On failure, this function returns the error
+// encoding status and error message string to describe the error, |emit|
+// function will not be called on failure. Only 32 and 64 bit floating point
+// numbers are supported.
+std::tuple<EncodeNumberStatus, std::string> ParseAndEncodeFloatingPointNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+// Parses an integer or floating point number of a given |type| from the given
+// |text| and encodes the number by the given |emit| function. On success,
+// returns EncodeNumberStatus::kSuccess and an empty error message string, the
+// parsed number will be consumed by the given |emit| function word by word
+// (least significant word first). On failure, this function returns the error
+// encoding status and error message string to describe the error, |emit|
+// function will not be called on failure. Integers up to 64 bits and 32/64
+// bit floating point values are supported.
+std::tuple<EncodeNumberStatus, std::string> ParseAndEncodeNumber(
+    const char* text, const IdType& type, std::function<void(uint32_t)> emit);
+
+}  // namespace spvutils
+
+// Definition of template functions.
+#include "util/parse_number_internal.h"
```

Comment:
This can't just be internal.  The templates have to apply to the parameterized types.  Which means the client needs to see what methods properties and operations can be applied to those types.  So this really has to be public, and seen as so.
For example, how do you know you can supply a FloatProxy object to these parsers?